File: src/main/java/dev/jsinco/luma/lumaevents/explorer/events/AbstractExplorerListener.java
package dev.jsinco.luma.lumaevents.explorer.events;

import dev.jsinco.luma.lumaevents.EventMain;
import dev.jsinco.luma.lumaevents.EventPlayerManager;
import dev.jsinco.luma.lumaevents.npc.constants.TutorialSection;
import dev.jsinco.luma.lumaevents.obj.EventPlayer;
import org.bukkit.Bukkit;
import org.bukkit.entity.Player;
import org.bukkit.event.Listener;

import java.util.UUID;

public abstract class AbstractExplorerListener implements Listener {

    public static void fire(Object event, Player player) {
        fire(event, player.getUniqueId());
    }

    public static void fireLater(Object event, Player player, long delay) {
        fireLater(event, player.getUniqueId(), delay);
    }

    public static void fire(Object event, UUID playerUUID) {
        EventPlayer eventPlayer = EventPlayerManager.getByUUID(playerUUID);
        if (!eventPlayer.hasCompletedTutorialSection(TutorialSection.EXPLORER_MILES)) {
            return;
        }

        Bukkit.getScheduler().runTaskAsynchronously(EventMain.getInstance(),
                () -> eventPlayer.fireForExplorerMiles(event)
        );
    }

    public static void fireLater(Object event, UUID playerUUID, long delay) {
        EventPlayer eventPlayer = EventPlayerManager.getByUUID(playerUUID);
        if (!eventPlayer.hasCompletedTutorialSection(TutorialSection.EXPLORER_MILES)) {
            return;
        }
        Bukkit.getScheduler().runTaskLaterAsynchronously(EventMain.getInstance(), () -> {
            if (!eventPlayer.isOnline()) {
                return; // Player logged off
            }
            eventPlayer.fireForExplorerMiles(event);
        }, delay);
    }
}

File: src/main/java/dev/lumas/events/explorer/listener/AbstractExplorerListener.kt
package dev.lumas.events.explorer.listener

import dev.lumas.events.explorer.intention.ExplorerIntentRegistry
import dev.lumas.events.manager.EventPlayerManager
import dev.lumas.events.model.EventPlayer
import dev.lumas.events.utility.Executors
import org.bukkit.entity.Entity
import org.bukkit.entity.Player
import org.bukkit.event.Listener
import java.util.concurrent.TimeUnit

interface AbstractExplorerListener : Listener {
    fun fire(event: Any, entity: Entity, ignoreMiles: Boolean = false) {
        if (entity is Player) {
            val eventPlayer: EventPlayer = EventPlayerManager.getByUUID(entity.uniqueId)
            if (eventPlayer.isSuspended) {
                eventPlayer.fireForExplorerOrders(entity.world, event)
            }
        }

        if (entity is Player) {
            val eventPlayer: EventPlayer = EventPlayerManager.getByUUID(entity.uniqueId)
            if (eventPlayer.isSuspended) {
                ExplorerIntentRegistry.unifiedValues().forEach { intent ->
                    intent.tryApply(entity.world, event)
                }
            }
        } else {
            ExplorerIntentRegistry.unifiedValues().forEach { intent ->
                intent.tryApply(entity.world, event)
            }
        }



        if (entity is Player && !ignoreMiles) {
            Executors.runAsync { _ ->
                val eventPlayer: EventPlayer = EventPlayerManager.getByUUID(entity.uniqueId)
                if (!eventPlayer.isSuspended) {
                    eventPlayer.fireForExplorerMiles(event)
                }
            }
        }
    }

    fun fireLater(event: Any, entity: Entity, delay: Long, ignoreMiles: Boolean = false) {

        if (entity is Player) {
            val eventPlayer: EventPlayer = EventPlayerManager.getByUUID(entity.uniqueId)

            Executors.delayedSync(entity, delay) {
                if (eventPlayer.isOnline && eventPlayer.isSuspended) {
                    eventPlayer.fireForExplorerOrders(entity.world, event)
                }
            }
        }

        Executors.delayedGlobal(delay) {
            Executors.runSync(entity) {
                ExplorerIntentRegistry.unifiedValues().forEach { intent ->
                    intent.tryApply(entity.world, event)
                }
            }
        }

        if (entity is Player && !ignoreMiles) {
            Executors.runDelayedAsync(TimeUnit.MILLISECONDS, delay * 50L) {
                val eventPlayer: EventPlayer = EventPlayerManager.getByUUID(entity.uniqueId)
                if (eventPlayer.isOnline && !eventPlayer.isSuspended) {
                    eventPlayer.fireForExplorerMiles(event)
                }
            }
        }
    }
}